decision_service/rlclientlib/data_callback_fn.h

```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include "model_mgmt.h"
+namespace reinforcement_learning { namespace model_management {
+
+  class data_callback_fn {
+  public:
+    using data_fn = void(*)(const model_data& data, void*);
+    int report_data(const model_data& data, api_status* status = nullptr);
+
+    data_callback_fn(data_fn, void*);
+    ~data_callback_fn() = default;
+
+    data_callback_fn(const data_callback_fn&) = delete;
+    data_callback_fn(data_callback_fn&&) = delete;
+    data_callback_fn& operator=(const data_callback_fn&) = delete;
+    data_callback_fn& operator=(data_callback_fn&&) = delete;
+
+  private:
+    data_fn _fn;
+    void* _context;
```

Comment:
>void* _context; [](start = 4, length = 15)

should we use shared_ptr to make lifespan management of the context easier?

Reply:
This is provider by the user of the class and returned to the user during callback.  Not owned by the callback.

---
In reply to: [193485951](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193485951) [](ancestors = 193485951)